nselib/smb2.lua

```diff
@@ -0,0 +1,410 @@
+  local ClientStartTime = overrides["ClientStartTime"] or 0 -- ClientStartTime only used in dialects > 3.11
+  local total_data = 0  -- Data counter
+  local padding_data = "" -- Padding string to align contexts
+  local context_data -- Holds Context data 
+  local is_0311 = false -- Flag for SMB 3.11 
+  local status, err
+
+  header = smb2_encode_header_sync(smb, command_codes['SMB2_COM_NEGOTIATE'], overrides)
+
+  -- We construct the first block that works for dialects 2.02 up to 3.11.
+  data = string.pack("<I2 I2 I2 I2 I4 c16", 
+    StructureSize,  -- 2 bytes: StructureSize
+    DialectCount,   -- 2 bytes: DialectCount
+    SecurityMode,   -- 2 bytes: SecurityMode
+    0,              -- 2 bytes: Reserved - Must be 0
+    Capabilities,   -- 4 bytes: Capabilities - 0 for dialects > 3.x
+    GUID            -- 16 bytes: ClientGuid
+  )
+
+  -- The next block gets interpreted in different ways depending on the dialect
+  if stdnse.contains(overrides['Dialects'], 0x0311) then
+    is_0311 = true
+  end
+
+  -- If we are dealing with 3.11 we need to set the following fields:
+  -- NegotiateContextOffset, NegotiateContextCount, and Reserved2
+  if is_0311 then
+    total_data = #header + #data + (DialectCount*2)
+    while ((total_data)%8 ~= 0) do
+        total_data = total_data + 1
+        padding_data = padding_data .. string.pack("<c1", 0x0)
+    end -- while to create 8 byte aligned padding
+    data = data .. string.pack("<I4 I2 I2", 
+                    total_data+8,   -- NegotiateContextOffset (4 bytes)
+                    0x2,            -- NegotiateContextCount (2 bytes) 
+                    0x0             -- Reserved2 (2 bytes)
+                   )
+  else  -- If it's not 3.11, the bytes are the ClientStartTime (8 bytes)
+    data = data .. string.pack("<I8", ClientStartTime) -- If it is not 3.11, we set it to 0
+  end -- if is_0311
+
+  -- Now we build the Dialect list, 16 bit integers
+  if(overrides['Dialects'] == nil) then  -- If no custom dialect is defined, used the default 2.10
+    data = data .. string.pack("<I2", 0x0210)
+  else  -- Dialects are set in overrides table
+    for _, v in ipairs(overrides['Dialects']) do
+      data = data .. string.pack("<I2", v)
+    end
+  end
+
+  -- If 3.11, we now need to add some padding between the dialects and the NegotiateContextList
+  -- I was only able to get this to work using both NegotiateContexts: 
+  -- * SMB2_PREAUTH_INTEGRITY_CAPABILITIES
+  -- * SMB2_ENCRYPTION_CAPABILITIES
+  if is_0311 then
+    data = data .. padding_data
+    local negotiate_context_list, context_data
+
+    -- We set SMB2_ENCRYPTION_CAPABILITIES first
+    context_data = string.pack("<I2 I2 I2",
+                    0x2,      -- CipherCount (2 bytes): 2 ciphers available
+                    0x0002,   -- Ciphers (2 bytes each): AES-128-GCM
+                    0x0001    -- Ciphers (2 bytes each): AES-128-CCM
+                  )
+    data = data .. string.pack("<I2 I2 I4 c" .. #context_data, 
```

Comment:
Instead of `string.pack("XYZ c" .. #something, X, Y, Z, something)`, just do `string.pack("XYZ", X, Y, Z) .. something`; it's simpler.